src/peer_manager.rs

```diff
@@ -296,17 +311,184 @@ impl PeerManager {
+            if name == candidate_name {
+                let result = if validity {
+                    let (pub_id, peer_id) = if let Some(peer) = self.peer_map
+                        .get_by_name(&candidate_name) {
+                        if let Some(peer_id) = peer.peer_id().clone() {
+                            (*peer.pub_id(), *peer_id)
+                        } else {
+                            return (false, None);
+                        }
+                    } else {
+                        return (false, None);
+                    };
+                    Some((pub_id, peer_id))
+                } else {
+                    None
+                };
+                self.node_candidate = None;
+                return (true, result);
+            }
+        }
+        (false, None)
+    }
+
+    pub fn handle_node_approval(&mut self) -> Vec<(PublicId, PeerId)> {
+        let mut result = vec![];
+        for peer_name in &self.peer_candidates.clone() {
+            if let Some(peer) = self.peer_map.get_by_name(peer_name) {
+                if let Some(peer_id) = peer.peer_id().clone() {
+                    result.push((*peer.pub_id(), *peer_id));
+                }
+            }
+        }
+        self.peer_candidates.clear();
+        result
+    }
+
+    pub fn is_node_candidate(&mut self,
+                             pub_id: &PublicId,
+                             peer_id: &PeerId)
+                             -> Result<Option<(bool, Vec<u8>)>, RoutingTableError> {
+        if let Some((ref name, _, ref seed)) = self.node_candidate {
+            if name == pub_id.name() {
+                let tunnel = match self.peer_map.remove(peer_id).map(|peer| peer.state) {
+                    Some(PeerState::SearchingForTunnel) |
+                    Some(PeerState::AwaitingNodeIdentify(true)) => true,
+                    Some(PeerState::Routing(_tunnel)) => {
+                        error!("NodeCandidate {:?} already in state Routing.", peer_id);
+                        return Err(RoutingTableError::AlreadyExists);
+                    }
+                    _ => false,
+                };
+                let state = PeerState::Candidate(tunnel);
+                let _ = self.peer_map.insert(Peer::new(*pub_id, Some(*peer_id), state));
+                return Ok(Some((tunnel, seed.clone())));
+            }
+        }
+        Ok(None)
+    }
+
+    pub fn is_peer_candidate(&mut self, pub_id: &PublicId, peer_id: &PeerId) -> bool {
+        if !self.peer_candidates.contains(pub_id.name()) {
+            return false;
+        }
+        let tunnel = match self.peer_map.remove(peer_id).map(|peer| peer.state) {
+            Some(PeerState::SearchingForTunnel) |
+            Some(PeerState::AwaitingNodeIdentify(true)) => true,
+            Some(PeerState::Routing(_tunnel)) => {
+                error!("PeerCandidate {:?} already in state Routing.", peer_id);
+                return true;
```

Comment:
If the function returns here the peer gets removed from the `peer_map` without replacement. I don't think this was intended?